src/transaction/meta_transaction.cpp

```diff
@@ -24,13 +24,28 @@ Transaction &Transaction::Get(ClientContext &context, AttachedDatabase &db) {
 	return meta_transaction.GetTransaction(db);
 }
 
+#ifdef DEBUG
+static void VerifyAllTransactionsUnique(AttachedDatabase &db, vector<reference<AttachedDatabase>> &all_transactions) {
+	for (auto &tx : all_transactions) {
+		if (RefersToSameObject(db, tx.get())) {
+			throw InternalException("Database is already present in all_transactions");
+		}
+	}
+}
+#endif
+
 Transaction &MetaTransaction::GetTransaction(AttachedDatabase &db) {
+	lock_guard<mutex> guard(lock);
 	auto entry = transactions.find(db);
```

Comment:
This lock is most likely only necessary in `MetaTransaction::GetTransaction` - Commit, Rollback and RemoveTransaction are all called in single-threaded mode under the transaction lock of the main transaction manager

Reply:
Regarding the reproduction, do you have a suggestion what I could try still?
I thought the only requirement would be to get the source to scan in parallel, so we get multiple worker threads getting MetaTransaction::Get at the same time

Reply:
Is the reproduction from the issue not good enough?

Reply:
I'm still confused by how this goes wrong in the first place - in this instance the transaction should be started as part of `bind_insert.cpp` when it binds the base table, and binding always happens single-threaded. I'm not sure if the parallel ingestion is the issue. Looking at the query I would suspect the sequence default value to be the culprit here.

Reply:
That's why I'm trying to create a better reproduction
all my hunches as to why this dies have been wrong so far

It seems to happen both with BatchedInsert and regular Insert

Reply:
> Looking at the query I would suspect the sequence default value to be the culprit here.

Yea that is the problem, because that calls NextValFunction - which calls MetaTransaction::Get
But I still can't reproduce this outside of the pandas scan

Reply:
This one seems to work for me:
```sql
CREATE SEQUENCE id_seq;
CREATE TABLE my_table (
    id INTEGER DEFAULT nextval('id_seq'),
    a INTEGER
);
ATTACH ':memory:' AS s1;
CREATE TABLE s1.tbl AS FROM range(2000000);                            ^
INSERT INTO my_table (a) SELECT * FROM s1.tbl;
INSERT INTO my_table (a) SELECT * FROM s1.tbl;
INSERT INTO my_table (a) SELECT * FROM s1.tbl;
INSERT INTO my_table (a) SELECT * FROM s1.tbl;
INSERT INTO my_table (a) SELECT * FROM s1.tbl;
INSERT INTO my_table (a) SELECT * FROM s1.tbl;
INSERT INTO my_table (a) SELECT * FROM s1.tbl;

```

Reply:
Transactions are launched within a database, the second database ensures we don't open a transaction as part of the scan.